api/src/main/java/org/apache/iceberg/variants/VariantDataUtil.java

```diff
@@ -0,0 +1,189 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.variants;
+
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.ObjectInputStream;
+import java.io.ObjectOutputStream;
+import java.math.BigDecimal;
+import java.nio.ByteBuffer;
+import java.util.List;
+import java.util.Map;
+import org.apache.iceberg.exceptions.RuntimeIOException;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.relocated.com.google.common.base.Splitter;
+import org.apache.iceberg.relocated.com.google.common.collect.ImmutableMap;
+import org.apache.iceberg.types.Type;
+import org.apache.iceberg.types.Types;
+
+public class VariantDataUtil {
+  private VariantDataUtil() {}
+
+  private static final Splitter DOT = Splitter.on(".");
+  private static final String ROOT = "$";
+
+  /**
+   * Extracts a nested value from nested variant objects, or returns null of no value matches.
+   *
+   * @param value a variant value
+   * @param fields a list of field names to extract
+   * @return the nested value identified by the fields
+   */
+  public static VariantValue extract(VariantValue value, List<String> fields) {
+    VariantValue current = value;
+    for (String field : fields) {
+      if (current != null && current.type() == PhysicalType.OBJECT) {
+        current = current.asObject().get(field);
+      } else {
+        return null;
+      }
+    }
+
+    return current;
+  }
+
+  public static ByteBuffer serializeBounds(Map<String, VariantValue> bounds) {
```

Comment:
This serialization is temporary and will be replaced with a well-defined variant serialization.